### DISCUSSION TITLE
Instructor does not support multiple tool calls

### DISCUSSION THREAD

**Discussion by agamm** [567-labs/instructor#892] [General]
- [X] This is actually a bug report.
- [ ] I am not getting good LLM Results
- [ ] I have tried asking for help in the community on discord or discussions and have not received a response.
- [ ] I have tried searching the documentation and have not found an answer.

**What Model are you using?**

- [ ] gpt-3.5-turbo
- [ ] gpt-4-turbo
- [ ] gpt-4
- [X] Other (please specify): Qwen2-72B-Instruct with litellm

**Describe the bug**
When having an unrelated response the structured output fails with:
> AssertionError: Instructor does not support multiple tool calls, use List[Model] instead.

**To Reproduce**
https://github.com/BerriAI/litellm/issues/4699

**Expected behavior**
Return an empty response and a warning.

**Screenshots**

**Comment by akarshghale** [comment]
Getting the same error with together ai async client when using the new llama 3.1 7B instruct.

**Comment by akarshghale** [comment]
Any updates on this?

**Comment by ssonal** [comment]
I can see that the deepinfra issue from [here](https://github.com/jxnl/instructor/issues/840?notification_referrer_id=NT_kwDOAB4c3rMxMTUwMzE5NjYyNToxOTczNDcw#issue-2407502614) and the llama 3.1B issue from [here](https://github.com/jxnl/instructor/issues/840?notification_referrer_id=NT_kwDOAB4c3rMxMTUwMzE5NjYyNToxOTczNDcw#issuecomment-2247143176) are syntax related issues because both providers are deviating from openAI compatible function calling. 

Deepinfra only [supports](https://deepinfra.com/docs/advanced/function_calling) `"tool_choice" = 'auto'` and Together.ai has its own unique [requirement](https://docs.together.ai/docs/llama-3-function-calling) for Llama 3.1 function calling.

**Comment by akarshghale** [comment]
> I can see that the deepinfra issue from [here](https://github.com/jxnl/instructor/issues/840?notification_referrer_id=NT_kwDOAB4c3rMxMTUwMzE5NjYyNToxOTczNDcw#issue-2407502614) and the llama 3.1B issue from [here](https://github.com/jxnl/instructor/issues/840?notification_referrer_id=NT_kwDOAB4c3rMxMTUwMzE5NjYyNToxOTczNDcw#issuecomment-2247143176) are syntax related issues because both providers are deviating from openAI compatible function calling. 
> 
> Deepinfra only [supports](https://deepinfra.com/docs/advanced/function_calling) `"tool_choice" = 'auto'` and Together.ai has its own unique [requirement](https://docs.together.ai/docs/llama-3-function-calling) for Llama 3.1 function calling.
> 

So what do you think can be done about it? 

There's no workaround other than updating the lib?

**Comment by akarshghale** [comment]
I think function calling and JSON mode are 2 different things in together ai right?

https://docs.together.ai/docs/json-mode

How can I use json mode instead of tool mode?

**Comment by akarshghale** [comment]
Yeah so using JSON_SCHEMA mode works just fine.

**Comment by agamm** [comment]
I'm not using together.ai, why do you bring it up?

**Comment by jxnl** [comment]
https://python.useinstructor.com/concepts/parallel/?h=parallel#understanding-parallel-function-calling

**Reply by agamm** [reply]
Can you briefly explain why is this related?

**Comment by AT-main** [comment]
I was getting the same error when using _Intstructor_ with a _llamafile model_. I ended up with the following script that worked for me.

```python
from instructor import Mode
import instructor
from pydantic import BaseModel
from openai import OpenAI


# Define your desired output structure
class UserInfo(BaseModel):
    name: str
    age: int


# Patch the OpenAI client
client = instructor.patch(OpenAI(base_url="http://localhost:8080/v1", api_key="not-needed"), mode=Mode.JSON_SCHEMA)

# Extract structured data from natural language
user_info = client.chat.completions.create(
    model="google_gemma-3-4b-it-Q6_K",
    response_model=UserInfo,
    messages=[{"role": "user", "content": "John Doe is 30 years old."}],
)

print(user_info.name)
#> John Doe
print(user_info.age)
#> 30
```
